src/control/security/signature.go

```diff
@@ -0,0 +1,108 @@
+//
+// (C) Copyright 2019 Intel Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+package security
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha512"
+	"io"
+
+	"github.com/pkg/errors"
+)
+
+//UnsupportedKeyError is a structured error used to indicate that the PublicKey
+//or PrivateKey interface passed in represents a key type we do not support.
+type UnsupportedKeyError struct{}
+
+//Error is the implementation of the error interface.
+func (err *UnsupportedKeyError) Error() string {
+	return "key contains an unsupported key type"
+}
+
+//TokenSigner serves to encapsulate the functionality needed
+//to sign and verify auth token signatures.
+type TokenSigner struct {
+	randPool io.Reader
+}
+
+//DefaultTokenSigner creates a TokenSigner with an instantiated entropy pool.
+func DefaultTokenSigner() *TokenSigner {
+	return &TokenSigner{
+		randPool: rand.Reader,
+	}
+}
+
+//Sign takes an unhashed set of bytes and hashes and signs the result with the
+//key passed in. If no key is specified it will return the unsigned digest.
```

Comment:
I don't understand the resistance to avoiding ambiguous behavior in API surface. It makes no sense to me that we have this thing which purports to sign things, but only if it's been given a non-nil key for signing, otherwise it returns something else.

Reply:
If I don't do this then I have to sprinkle key checks and insecure flag checks in a bunch of other places.

Reply:
Or pass additional information down the call stack.

Reply:
That would be preferable to this. This is hard to reason about (as evidenced by my confusion when I saw this called elsewhere). Explicit is better than clever. Particularly when we're dealing with security-focused code.

Reply:
fixed